tools/gcs-usage-recommender/python/backfill.py

```diff
@@ -0,0 +1,124 @@
+# Copyright 2019 Google Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import datetime
+import json
+import sys
+import logging
+from typing import Dict, List
+from google.cloud import resource_manager
+from google.cloud import storage
+
+
+def get_storage_client() -> storage.Client:
+    """Retrieves GCS client.
+
+    Returns:
+        storage.Client object
+    """
+    return storage.Client()
+
+
+def get_resource_manager_client() -> resource_manager.Client:
+    """Retrieves Resource Manager Cleint
+
+    Returns:
+        resource_manager.Client object
+    """
+    return resource_manager.Client()
+
+
+def get_project_ids(resource_mgr_client: resource_manager.Client,
+                    organization_id: int) -> List[str]:
+    """Generates list of project_ids within a GCP org.
+
+    Args:
+        resource_mgr_client: resource_manager.Client
+        organization_id: Integer representing GCP org ID
+
+    Returns:
+        List of strings holding all project IDs.
+    """
+    project_id_list = []
+    project_filter = {
+        'parent.id': organization_id,
+        'parent.type': 'organization'
+    }
+    for project in resource_mgr_client.list_projects(project_filter):
+        project_id_list.append(project.project_id)
+    return project_id_list
+
+
+def get_buckets(project_ids: List[str],
+                gcs_client: storage.Client) -> List[Dict[str, str]]:
+    """Retrieves list of metadata for all buckets in a GCP org.
+
+    Args:
+        project_ids: List of strings holding project IDs
+        gcs_client: storage.Client object
+
+    Returns:
+        List of dictionaries mapping bucket-level metadata.
+    """
+    output_list = []
+    try:
+        for project_id in project_ids:
+            bucket_list = list(gcs_client.list_buckets(project=project_id))
+            for bucket in bucket_list:
+                output_list.append({
```

Comment:
I'm having trouble here. Can you help me understand the significance of the values in this JSON?
Are these expected to be overwritten by BQ later?
Would it be safer to set las_read_timestamp to 30 days ago to avoid suggesting over aggressive OLM Policies?

Reply:
Yes, these are meant to capture all buckets that existed in an org prior to enabling audit logs. Because of this, they have default dummy values, that are overwritten once the bucket appears in the audit logs.

Open to suggestions on last_read_timestamp, although I would think that a more obvious dummy value would perhaps be better to show that it truly has not been accessed. A 30-days-ago timestamp may lead someone to incorrectly believed that it was read, when in fact, it maybe hasn't been accessed in a year. In this case, the OLM is blank regardless, so I don't think aggressive OLM policies will be in effect.

Reply:
This is a tricky one and you're right to have concern about misleading the user on last read date. I'd argue that the same could be said about setting the dummy date so far back misleading them to believe an object hasn't been read for years when perhaps it has been read in the last 90 days.
If we set 30 days ago date we'll recommend near line for all the backfill data (that hasn't appeared in the audit logs). Then after 60 days we'll either recommend going to cold line or not based on actual information we observed.   
I think we really want to avoid a scenario where we suggest moving many objects to cold line storage and then have to convert many of them back. Changing storage class of many objects can take GCS a long time and affects read performance. This can be very dramatic for customers who are moving stuff to cold line and then simultaneously trying to run a yearly report on those same objects.

Curious what you'd think about a null date?
It might be annoying to deal with but I think it avoids this pickle of picking a misleading default.

Reply:
Hmm.. agreed on the performance of changing storage classes. However, this defaults OLM to be blank, btw, not coldline or nearline, so whichever dummy value we decide on affects just a user's understanding of the bucket not an OLM. The OLM is generated only when it appears in the audit logs.

1. My only concern about doing a null date was because I wanted the schema to default to datetime type, but I just realized that I already set up the schema file. I agree and will change to null :)